### PR TITLE
Make network interface reconcile/delete async

### LINKED ISSUE #1717: async networkinterfaces

### PR DESCRIPTION
**Pull request by Jont828**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it**: Implementation of an async service for network interfaces as part of an effort to make all services async. See #1610 and #1541.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1717 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Make network interfaces reconcile/delete async
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
comment from @shysank about using named returns from another PR that is relevant here:
#1686 (comment)

Here is the commit that does this in the route tables PR: b2302ea

File: azure/services/networkinterfaces/client.go [hunk 1]
```diff
@@ -53,44 +62,114 @@ func newInterfacesClient(subscriptionID string, baseURI string, authorizer autor
 	return nicClient
 }
 
-// Get gets information about the specified network interface.
-func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, nicName string) (network.Interface, error) {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "networkinterfaces.AzureClient.Get")
+// Get gets the specified network interface.
+func (ac *AzureClient) Get(ctx context.Context, spec azure.ResourceSpecGetter) (result interface{}, err error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "interfaces.AzureClient.Get")
```
Comment (mboersma) [inline azure/services/networkinterfaces/client.go:67]: Shouldn't the logger span still be labeled with `networkinterfaces`? I thought it was always "packagename.struct.func".

File: azure/services/networkinterfaces/client.go [hunk 2]
```diff
@@ -53,44 +62,114 @@ func newInterfacesClient(subscriptionID string, baseURI string, authorizer autor
...
 }
 
-// CreateOrUpdate creates or updates a network interface.
-func (ac *AzureClient) CreateOrUpdate(ctx context.Context, resourceGroupName string, nicName string, nic network.Interface) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "networkinterfaces.AzureClient.CreateOrUpdate")
+// CreateOrUpdateAsync creates or updates a network interface asynchronously.
+// It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (result interface{}, future azureautorest.FutureAPI, err error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "interfaces.AzureClient.CreateOrUpdateAsync")
```
Comment (mboersma) [inline azure/services/networkinterfaces/client.go:77]: Same question about logger span label here, and elsewhere.

**Review by mboersma — COMMENTED** [review]
This is looking good! There is still a comment in `machine_test.go` about adding another test case:
`// TODO: add a test for when the cache is not empty`

File: azure/services/networkinterfaces/client.go (hunk 1 above)
Comment (Jont828) [reply azure/services/networkinterfaces/client.go]: Yeah it should, thanks for catching that!

File: azure/scope/machine.go [hunk 3]
```diff
@@ -216,18 +223,22 @@ func (m *MachineScope) InboundNatSpecs(portsInUse map[int32]struct{}) []azure.Re
...
+		SubscriptionID:        m.SubscriptionID(),
 		MachineName:           m.Name(),
 		VNetName:              m.Vnet().Name,
 		VNetResourceGroup:     m.Vnet().ResourceGroup,
 		SubnetName:            m.AzureMachine.Spec.SubnetName,
-		VMSize:                m.AzureMachine.Spec.VMSize,
 		AcceleratedNetworking: m.AzureMachine.Spec.AcceleratedNetworking,
 		IPv6Enabled:           m.IsIPv6Enabled(),
 		EnableIPForwarding:    m.AzureMachine.Spec.EnableIPForwarding,
+		SKU:                   nil,
```
Comment (CecileRobertMichon) [inline azure/scope/machine.go:239]: this shouldn't be needed

```suggestion
```

File: azure/scope/machine.go [hunk 4]
```diff
@@ -250,15 +261,19 @@ func (m *MachineScope) NICSpecs() []azure.NICSpec {
...
+	return []azure.ResourceSpecGetter{spec}
 }
 
 // NICIDs returns the NIC resource IDs.
 func (m *MachineScope) NICIDs() []string {
 	nicspecs := m.NICSpecs()
 	nicIDs := make([]string, len(nicspecs))
 	for i, nic := range nicspecs {
-		nicIDs[i] = azure.NetworkInterfaceID(m.SubscriptionID(), m.ResourceGroup(), nic.Name)
+		nicIDs[i] = azure.NetworkInterfaceID(m.SubscriptionID(), m.ResourceGroup(), nic.ResourceName())
```
Comment (CecileRobertMichon) [inline azure/scope/machine.go:276]:
```suggestion
		nicIDs[i] = azure.NetworkInterfaceID(m.SubscriptionID(), nic.ResourceGroupName(), nic.ResourceName())
```

should be equivalent but this is slightly more precise, and more resilient if we ever support having the NIC in a separate resource group

File: azure/services/networkinterfaces/networkinterfaces.go [hunk 5]
```diff
@@ -19,158 +19,86 @@ package networkinterfaces
...
 	azure.ClusterDescriber
-	NICSpecs() []azure.NICSpec
+	azure.AsyncStatusUpdater
+	NICSpecs() []azure.ResourceSpecGetter
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
 	Scope NICScope
 	Client
```
Comment (CecileRobertMichon) [inline azure/services/networkinterfaces/networkinterfaces.go:42]: is `Client` still being used anywhere?

Comment (Jont828) [reply azure/services/networkinterfaces/networkinterfaces.go:42]: Yep, it looks like it's being used in the VM service.

Comment (CecileRobertMichon) [reply azure/services/networkinterfaces/networkinterfaces.go:42]: not the Client interface, the Client in the Service struct. I don't think you're using this Client anywhere in the networkinterfaces service.

File: azure/scope/machine.go [hunk 6]
```diff
@@ -136,6 +138,11 @@ func (m *MachineScope) InitMachineCache(ctx context.Context) error {
 		if err != nil {
 			return errors.Wrapf(err, "failed to get availability set SKU %s in compute api", string(compute.AvailabilitySetSkuTypesAligned))
 		}
+
+		m.cache.networkInterfaceSKU, err = skuCache.Get(ctx, m.AzureMachine.Spec.VMSize, resourceskus.VirtualMachines)
```
Comment (CecileRobertMichon) [inline azure/scope/machine.go:142]: This seems to be the same as the VMSKU above? If so can we reuse it?

File: api/v1beta1/conditions_consts.go [hunk 7]
```diff
@@ -116,6 +116,8 @@ const (
 	RoleAssignmentReadyCondition clusterv1.ConditionType = "RoleAssignmentReady"
 	// DisksReadyCondition means the disks exist and are ready to be used.
 	DisksReadyCondition clusterv1.ConditionType = "DisksReady"
+	// NetworkInterfaceReadyCondition means the network interfaces exist and are ready to be used.
+	NetworkInterfaceReadyCondition clusterv1.ConditionType = "NetworkInterfacesReady"
```
Comment (CecileRobertMichon) [inline api/v1beta1/conditions_consts.go:120]: also needs to be added to PatchObject summary/owned conditions in machine.go

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
